File: test/scripts/benchmark_db_api_client.py
#!/usr/bin/env python
"""
Benchmark Database API Client

This module provides a Python client for the Benchmark Database API.
"""

import os
import sys
import json
import logging
import datetime
import requests
from typing import Dict, List, Any, Optional, Union
from pathlib import Path
import duckdb

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger("benchmark_db_client")

class BenchmarkDBAPI:
    """Client API for the benchmark database"""
    
    def __init__(self, database_path="./benchmark_db.duckdb", server_url=None):
        """
        Initialize the database API.
        
        Args:
            database_path: Path to the DuckDB database file
            server_url: URL of the database API server (if using server mode)
        """
        self.database_path = database_path
        self.server_url = server_url
        self.conn = None
        
        # If using direct database access, create required tables if they don't exist
        if server_url is None:
            # Ensure the database directory exists
            os.makedirs(os.path.dirname(os.path.abspath(database_path)), exist_ok=True)
            self.ensure_database_initialized()
    
    def get_connection(self):
        """Get a connection to the database"""
        if self.conn is None:
            try:
                # Check if database file exists, create it if it doesn't
                db_path = Path(self.database_path)
                if not db_path.exists():
                    # Create parent directory if it doesn't exist
                    db_path.parent.mkdir(parents=True, exist_ok=True)
                
                self.conn = duckdb.connect(self.database_path)
                
                # Initialize tables if they don't exist
                self.ensure_database_initialized()
            except Exception as e:
                logger.error(f"Error connecting to database: {e}")
                raise
        
        return self.conn
    
    def close(self):
        """Close the database connection"""
        if self.conn is not None:
            self.conn.close()
            self.conn = None
    
    def ensure_database_initialized(self):
        """Initialize the database if needed"""
        self.ensure_web_platform_table_exists()
    
    def ensure_web_platform_table_exists(self):
        """Ensure the web_platform_results table exists"""
        conn = self.get_connection()
        
        try:
            # Check if table exists
            table_exists = conn.execute("""
                SELECT name FROM sqlite_master 
                WHERE type='table' AND name='web_platform_results'
            """).fetchone()
            
            if not table_exists:
                # Create table if it doesn't exist
                conn.execute("""
                CREATE TABLE web_platform_results (
                    result_id INTEGER PRIMARY KEY,
                    model_name VARCHAR,
                    model_type VARCHAR,
                    browser VARCHAR,
                    platform VARCHAR,
                    status VARCHAR,
                    execution_time DOUBLE,
                    metrics JSON,
                    error_message VARCHAR,
                    source_file VARCHAR,
                    timestamp TIMESTAMP
                )
                """)
                logger.info("Created web_platform_results table")
                
                # Create indices
                conn.execute("CREATE INDEX idx_wpr_model_type ON web_platform_results(model_type)")
                conn.execute("CREATE INDEX idx_wpr_browser ON web_platform_results(browser)")
                conn.execute("CREATE INDEX idx_wpr_platform ON web_platform_results(platform)")
                conn.execute("CREATE INDEX idx_wpr_timestamp ON web_platform_results(timestamp)")
                
                return True
            
            return False
        except Exception as e:
            logger.error(f"Error ensuring web_platform_results table exists: {e}")
            raise
    
    def store_web_platform_result(self, 
                                model_name: str,
                                model_type: str,
                                browser: str,
                                platform: str,
                                status: str,
                                metrics: Optional[Dict[str, Any]] = None,
                                execution_time: Optional[float] = None,
                                error_message: Optional[str] = None,
                                source_file: Optional[str] = None,
                                timestamp: Optional[datetime.datetime] = None) -> int:
        """
        Store a web platform test result in the database.
        
        Args:
            model_name: Name of the model tested
            model_type: Type of model (whisper, wav2vec2, clap, etc.)
            browser: Browser used for testing (chrome, firefox, safari, edge)
            platform: Web platform tested (webnn, webgpu)
            status: Test status (successful, failed, etc.)
            metrics: Dictionary of test metrics (optional)
            execution_time: Execution time in seconds (optional)
            error_message: Error message if test failed (optional)
            source_file: Source file containing the test results (optional)
            timestamp: Timestamp for the test result (optional, defaults to now)
            
        Returns:
            ID of the stored result
        """
        # Use REST API if server URL is provided
        if self.server_url is not None:
            try:
                response = requests.post(
                    f"{self.server_url}/api/web-platform",
                    json={
                        "model_name": model_name,
                        "model_type": model_type,
                        "browser": browser,
                        "platform": platform,
                        "status": status,
                        "metrics": metrics,
                        "execution_time": execution_time,
                        "error_message": error_message,
                        "source_file": source_file,
                        "timestamp": timestamp.isoformat() if timestamp else datetime.datetime.now().isoformat()
                    }
                )
                response.raise_for_status()
                return response.json()["result_id"]
            except Exception as e:
                logger.error(f"Error storing web platform result via API: {e}")
                raise
        
        # Use direct database access
        conn = self.get_connection()
        
        # Ensure required table exists
        self.ensure_web_platform_table_exists()
        
        try:
            # Set default timestamp if not provided
            if timestamp is None:
                timestamp = datetime.datetime.now()
            
            # Convert metrics to JSON string if provided
            metrics_json = None
            if metrics is not None:
                metrics_json = json.dumps(metrics)
            
            # Insert into database
            result = conn.execute("""
                INSERT INTO web_platform_results (
                    model_name, model_type, browser, platform, status, 
                    execution_time, metrics, error_message, source_file, timestamp
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                RETURNING result_id
            """, (
                model_name, model_type, browser, platform, status,
                execution_time, metrics_json, error_message, source_file, timestamp
            )).fetchone()
            
            # Get the inserted ID
            result_id = result[0]
            
            logger.debug(f"Stored web platform result with ID: {result_id}")
            return result_id
        except Exception as e:
            logger.error(f"Error storing web platform result: {e}")
            raise
    
    def query_web_platform_results(self,
                                 model_name: Optional[str] = None,
                                 model_type: Optional[str] = None,
                                 browser: Optional[str] = None,
                                 platform: Optional[str] = None,
                                 status: Optional[str] = None,
                                 start_date: Optional[str] = None,
                                 end_date: Optional[str] = None,
                                 limit: int = 100) -> List[Dict[str, Any]]:
        """
        Query web platform test results from the database.
        
        Args:
            model_name: Filter by model name (optional)
            model_type: Filter by model type (optional)
            browser: Filter by browser (optional)
            platform: Filter by platform (optional)
            status: Filter by status (optional)
            start_date: Filter by start date (optional)
            end_date: Filter by end date (optional)
            limit: Maximum number of results to return (optional)
            
        Returns:
            List of web platform test results
        """
        # Use REST API if server URL is provided
        if self.server_url is not None:
            try:
                params = {
                    "limit": limit
                }
                
                if model_name:
                    params["model_name"] = model_name
                
                if model_type:
                    params["model_type"] = model_type
                
                if browser:
                    params["browser"] = browser
                
                if platform:
                    params["platform"] = platform
                
                if status:
                    params["status"] = status
                
                if start_date:
                    params["start_date"] = start_date
                
                if end_date:
                    params["end_date"] = end_date
                
                response = requests.get(
                    f"{self.server_url}/api/web-platform",
                    params=params
                )
                response.raise_for_status()
                return response.json()
            except Exception as e:
                logger.error(f"Error querying web platform results via API: {e}")
                raise
        
        # Use direct database access
        conn = self.get_connection()
        
        # Ensure table exists
        try:
            self.ensure_web_platform_table_exists()
        except Exception:
            # If table doesn't exist, return empty list
            return []
        
        try:
            # Build query with filters
            query = """
            SELECT 
                result_id,
                model_name,
                model_type,
                browser,
                platform,
                status,
                execution_time,
                metrics,
                error_message,
                source_file,
                timestamp
            FROM 
                web_platform_results
            """
            
            params = []
            where_clauses = []
            
            if model_name:
                where_clauses.append("model_name LIKE ?")
                params.append(f"%{model_name}%")
            
            if model_type:
                where_clauses.append("model_type = ?")
                params.append(model_type)
            
            if browser:
                where_clauses.append("browser = ?")
                params.append(browser)
            
            if platform:
                where_clauses.append("platform = ?")
                params.append(platform)
            
            if status:
                where_clauses.append("status = ?")
                params.append(status)
            
            if start_date:
                try:
                    start_dt = datetime.datetime.fromisoformat(start_date.replace('Z', '+00:00'))
                    where_clauses.append("timestamp >= ?")
                    params.append(start_dt)
                except ValueError as e:
                    logger.error(f"Invalid date format for start_date: {e}")
                    raise ValueError(f"Invalid date format for start_date: {start_date}")
            
            if end_date:
                try:
                    end_dt = datetime.datetime.fromisoformat(end_date.replace('Z', '+00:00'))
                    where_clauses.append("timestamp <= ?")
                    params.append(end_dt)
                except ValueError as e:
                    logger.error(f"Invalid date format for end_date: {e}")
                    raise ValueError(f"Invalid date format for end_date: {end_date}")
            
            if where_clauses:
                query += " WHERE " + " AND ".join(where_clauses)
            
            query += " ORDER BY timestamp DESC LIMIT ?"
            params.append(limit)
            
            # Execute query
            results = conn.execute(query, params).fetchdf()
            
            # Convert to list of dicts and parse metrics JSON
            results_list = []
            for _, row in results.iterrows():
                result = row.to_dict()
                
                # Parse metrics JSON
                if 'metrics' in result and result['metrics'] is not None:
                    try:
                        result['metrics'] = json.loads(result['metrics'])
                    except json.JSONDecodeError:
                        result['metrics'] = {}
                
                results_list.append(result)
            
            return results_list
        except Exception as e:
            logger.error(f"Error querying web platform results: {e}")
            raise

File: ipfs_accelerate_js/src/utils/benchmark_db_api_client.ts
// FI: any;
 * Convert: any;
 * Conversi: any;
 * Th: any;
 * Conversi: any;
 */;



// WebG: any;
/** Benchma: any;

Th: any;

impo: any;
impo: any;
impo: any;
impo: any;
impo: any;
impo: any;
impo: any;

// Configu: any;
logging.basicConfig() {);
level) { any) { any: any = loggi: any;
format: any: any: any: any: any: any = '%())asctime)s - %())name)s - %())levelname)s - %())message)s';'
);
logger: any: any: any = loggi: any;
;
class $1 extends $2 {/** Client API for (((((the benchmark database */}
  $1($2) {/** Initialize the database API.}
    Args) {
      database_path) { Path) { an) { an: any;
      server_url) { URL of the database API server ())if (((((using server mode) { */;
      this.database_path = database_pat) { an) { an: any;
      this.server_url = server_u) { an: any;
      this.conn = n: any;
    ;
    // If using direct database access, create required tables if (((($1) {
    if ($1) {
      // Ensure) { an) { an: any;
      os.makedirs())os.path.dirname())os.path.abspath())database_path)), exist_ok) { any) {any = tru) { an: any;
      th: any;
  $1($2) {
    /** G: any;
    if (((((($1) {
      try {
        // Check) { an) { an: any;
        db_path) { any) { any) { any = Path())this.database_path)) {
        if ((((((($1) { ${$1} catch(error) { any)) { any {logger.error())`$1`)}
          rais) { an) { an: any;
    
      }
          retur) { an: any;
  
    }
  $1($2) {
    /** Clo: any;
    if (((((($1) {this.conn.close());
      this.conn = nul) { an) { an: any;};
  $1($2) {
    /** Initializ) { an: any;
    th: any;
  ) {}
  $1($2) {
    /** Ensu: any;
    conn) {any = th: any;};
    try {
      // Che: any;
      table_exists) { any) { any: any = conn.execute() {)/** SELE: any;
      WHERE type: any: any = 'table' AND name: any: any: any: any: any: any = 'web_platform_results' */).fetchone());'
      ) {
      if ((((((($1) { ${$1} catch(error) { any)) { any {logger.error())`$1`)}
        rais) { an) { an: any;
  
    }
        functio) { an: any;
        $1) {string,;
        $1: stri: any;
        $1: stri: any;
        $1: stri: any;
        $1: stri: any;
        metrics: Record<str, Any | null> = nu: any;
        $1: $2 | null: any: any: any = nu: any;
        $1: $2 | null: any: any: any = nu: any;
        $1: $2 | null: any: any: any = nu: any;
        timestamp: datetime.datetime | null: any = nu: any;
        /** Sto: any;
      model_n: any;
      model_t: any;
      browser: Browser used for ((((((testing () {)chrome, firefox) { any) { an) { an: any;
      platform) { We) { an: any;
      status) { Te: any;
      metr: any;
      execution_t: any;
      error_message: Error message if ((((((($1) {
        source_file) { Source) { an) { an: any;
        timestamp) { Timestamp for ((((((the test result () {)optional, defaults to now)}
    Returns) {
      ID) { an) { an: any;
    // Use REST API if ((((($1) {) {
    if (($1) {
      try {
        response) { any) { any) { any) { any = requests) { an) { an: any;
        `$1`,;
        json) { any) { any: any = {}
        "model_name") { model_na: any;"
        "model_type": model_ty: any;"
        "browser": brows: any;"
        "platform": platfo: any;"
        "status": stat: any;"
        "metrics": metri: any;"
        "execution_time": execution_ti: any;"
        "error_message": error_messa: any;"
        "source_file": source_fi: any;"
        "timestamp": timestamp.isoformat()) if ((((((timestamp else { datetime.datetime.now() {).isoformat())}"
        );
        response) { an) { an: any;
      return response.json())["result_id"]) {} catch(error) { any)) { any {logger.error())`$1`);"
      rais) { an: any;
    }
      conn: any: any: any = th: any;
    
  }
    // Ensu: any;
    }
      th: any;
    ;
    try {
      // Set default timestamp if ((((((($1) {
      if ($1) {
        timestamp) {any = datetime) { an) { an: any;}
      // Conver) { an: any;
      };
      metrics_json) { any) { any: any = null) {
      if ((((((($1) { ${$1} catch(error) { any)) { any {logger.error())`$1`)}
        rais) { an) { an: any;
  
    }
        functio) { an: any;
        $1) { $2 | null: any: any: any = nu: any;
        $1: $2 | null: any: any: any = nu: any;
        $1: $2 | null: any: any: any = nu: any;
        $1: $2 | null: any: any: any = nu: any;
        $1: $2 | null: any: any: any = nu: any;
        $1: $2 | null: any: any: any = nu: any;
        $1: $2 | null: any: any: any = nu: any;
        $1: number: any: any = 1: any;
        /** Que: any;
    
    A: any;
      model_n: any;
      model_t: any;
      brow: any;
      platf: any;
      sta: any;
      start_d: any;
      end_d: any;
      li: any;
      
    Retu: any;
      Li: any;
    // Use REST API if ((((((($1) {) {
    if (($1) {
      try {
        params) { any) { any) { any) { any = {}
        "limit") {limit}"
        if (((((($1) {
          params["model_name"] = model_nam) { an) { an: any;"
          ,;
        if ((($1) {
          params["model_type"] = model_typ) { an) { an: any;"
          ,;
        if ((($1) {
          params["browser"] = browse) { an) { an: any;"
          ,;
        if ((($1) {
          params["platform"] = platfor) { an) { an: any;"
          ,;
        if ((($1) {
          params["status"] = statu) { an) { an: any;"
          ,;
        if ((($1) {
          params["start_date"] = start_dat) { an) { an: any;"
          ,;
        if ((($1) { ${$1} catch(error) { any)) { any {logger.error())`$1`)}
          rais) { an) { an: any;
    
        }
    // Us) { an: any;
        }
          conn: any: any: any = th: any;
    
        }
    // Ensu: any;
        };
    try ${$1} catch(error: any): any {
      // I: an: any;
      retu: any;
      ,;
    try {// Bui: any;
      query: any: any: any = /** SELE: any;
      result_: any;
      model_n: any;
      model_ty: any;
      brow: any;
      platfo: any;
      sta: any;
      execution_ti: any;
      metr: any;
      error_messa: any;
      source_f: any;
      timest: any;
      FROM 
      web_platform_results */}
      params: any: any: any: any: any: any = [],;
      ,    where_clauses: any: any: any: any: any: any = [],;
      ,;
      if (((((($1) {$1.push($2))"model_name LIKE) { an) { an: any;"
        $1.push($2))`$1`)}
      if ((($1) {$1.push($2))"model_type = ?");"
        $1.push($2))model_type)};
      if ($1) {$1.push($2))"browser = ?");"
        $1.push($2))browser)};
      if ($1) {$1.push($2))"platform = ?");"
        $1.push($2))platform)};
      if ($1) {$1.push($2))"status = ?");"
        $1.push($2))status)};
      if ($1) {
        try ${$1} catch(error) { any)) { any {logger.error())`$1`);
          throw new ValueError())`$1`)}
      if ((($1) {
        try ${$1} catch(error) { any)) { any {logger.error())`$1`);
          throw new ValueError())`$1`)}
      if ((($1) {query += " WHERE " + " AND ".join())where_clauses)}"
        query += " ORDER) { an) { an: any;"
        $1.push($2))limit);
      
      }
      // Execut) { an: any;
      }
        results) {any = conn.execute())query, params) { a: any;;}
      // Conve: any;
        }
        results_list: any: any: any: any: any: any = [],;
        };
    ,    for (((((_) { any, row in results.iterrows() {)) {}
      result) { any) { any) { any = ro) { an: any;
        
        // Par: any;
      if (((((($1) {,;
          try {
            result["metrics"] = json) { an) { an: any;"
          catch (error) { any) {}
            result["metrics"] = {}"
            ,;
            $1.push($2))result);
      
            retur) { an: any;
    } catch(error: any)) { any {
      log: any;
            r: any;